lib/maths/CDataFrameUtils.cc

```diff
@@ -117,6 +122,113 @@ bool CDataFrameUtils::columnQuantiles(std::size_t numberThreads,
     return true;
 }
 
+CDataFrameUtils::TDoubleVec CDataFrameUtils::micWithColumn(const core::CDataFrame& frame,
+                                                           const TSizeVec& columnMask,
+                                                           std::size_t targetColumn) {
+
+    using TFloatVec = std::vector<CFloatStorage>;
+    using TFloatVecVec = std::vector<TFloatVec>;
+    using TRowSampler = CSampling::CRandomStreamSampler<TRowRef>;
+
+    TDoubleVec mics(frame.numberColumns());
+
+    std::size_t n{std::min(NUMBER_SAMPLES_TO_COMPUTE_MIC, frame.numberRows())};
+
+    CPRNG::CXorOShiro128Plus rng;
+    TFloatVecVec samples;
+
+    if (frame.inMainMemory()) {
+
+        for (auto i : columnMask) {
+
+            auto onSample = [&](std::size_t slot, const TRowRef& row) {
+                if (slot >= samples.size()) {
+                    samples.resize(slot + 1, TFloatVec(2));
+                }
+                samples[slot][0] = row[i];
+                samples[slot][1] = row[targetColumn];
+            };
+            TRowSampler sampler{n, onSample};
+
+            auto result = frame.readRows(
+                1, core::bindRetrievableState(
+                       [&](std::size_t& missing, TRowItr beginRows, TRowItr endRows) {
+                           for (auto row = beginRows; row != endRows; ++row) {
+                               if (isMissing((*row)[i])) {
+                                   ++missing;
+                               } else if (isMissing((*row)[targetColumn]) == false) {
+                                   sampler.sample(*row);
+                               }
+                           }
+                       },
+                       std::size_t{0}));
+            LOG_TRACE(<< "# samples = " << samples.size());
+
+            double fractionMissing{static_cast<double>(result.first[0].s_FunctionState) /
+                                   static_cast<double>(frame.numberRows())};
+            LOG_TRACE(<< "Fraction missing = " << fractionMissing);
+
```

Comment:
The rationale here is that for this branch it is cheap to visit the data frame multiple times, so I maintain a set of samples for only the column `i`, and visit the data frame `|columnMask|` times. If the data frame is on disk (else branch) this is not the case and so I sample the full rows and only visit once. This makes it a little fiddly to share the code. Let me have a think if there is anything I can do to help.

Reply:
I decided in the end that rather than trying to share code I'd just move the two branches into their own functions. I think this improves readability. Trying to share sampling code (with callbacks) complicated matters to the point that I think it was worse. See [b078457](https://github.com/elastic/ml-cpp/pull/488/commits/b078457de24a98aa92ebd3559e7d67f9eb54c9f7).